apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -242,53 +240,79 @@ impl ApplyTo for NamedSelection {
-// note in selection_set.rs for more detail.
-pub(super) const TYPENAMES: &str = "$typenames";
-
 impl ApplyTo for PathSelection {
     fn apply_to_path(
         &self,
         data: &JSON,
-        vars: &IndexMap<String, JSON>,
-        input_path: &mut Vec<JSON>,
+        vars: &VarsWithPathsMap,
+        input_path: &InputPath<JSON>,
         errors: &mut IndexSet<ApplyToError>,
     ) -> Option<JSON> {
-        if let JSON::Array(array) = data {
-            return self.apply_to_array(array, vars, input_path, errors);
-        }
-
-        match self {
-            Self::Var(var_name, tail) => {
-                if var_name == "$" {
-                    // Because $ refers to the current value, we keep using
-                    // input_path instead of creating a new var_path here.
-                    tail.apply_to_path(data, vars, input_path, errors)
-                } else if var_name == TYPENAMES {
-                    if let PathSelection::Key(Key::Field(ref name), _) = **tail {
-                        let var_data = json!({ name: name });
-                        let mut var_path = vec![json!(name)];
-                        tail.apply_to_path(&var_data, vars, &mut var_path, errors)
+        match &self.path {
+            // If this is a KeyPath, instead of using data as given, we need to
+            // evaluate the path starting from the current value of $. To
+            // evaluate the KeyPath against data, prefix it with @. This logic
+            // supports method chaining like obj->has('a')->and(obj->has('b')),
+            // where both obj references are interpreted as $.obj.
+            PathList::Key(key, tail) => {
+                if let Some((dollar_data, dollar_path)) = vars.get("$") {
+                    let input_path_with_key = dollar_path.append(key.to_json());
+                    if let Some(child) = dollar_data.get(key.as_string()) {
+                        tail.apply_to_path(child, vars, &input_path_with_key, errors)
                     } else {
                         errors.insert(ApplyToError::new(
-                            format!("Invalid {} usage", TYPENAMES).as_str(),
-                            &[json!(var_name), json!(tail)],
+                            format!(
+                                "Property {} not found in {}",
+                                key.dotted(),
+                                json_type_name(dollar_data),
+                            )
+                            .as_str(),
+                            input_path_with_key.to_vec(),
                         ));
                         None
                     }
-                } else if let Some(var_data) = vars.get(var_name) {
-                    let mut var_path = vec![json!(var_name)];
-                    tail.apply_to_path(var_data, vars, &mut var_path, errors)
+                } else {
+                    // If $ is undefined for some reason, fall back to using data.
```

Comment:
This should be impossible... right? I wonder if we need `vars` to be a struct with all the defined things we _know_ exist, rather than a map 🤔

Reply:
Yes, impossible because we set `$` unconditionally in `apply_with_vars`.

Interesting idea about using a struct with known field names instead of dynamic map, but I'd like to see what other kinds of variables we end up wanting, before closing that door.

Reply:
I think that Dylan's suggestion here to use a struct is a good one, as we can catch missing implementations at compile time.

Reply:
The downside of using a struct is that lookup has to match `var_name` linearly against the known `&str` values, 
then repetitively perform the same operation against the corresponding struct fields. And the compiler can't tell if you're matching against all possible `var_name` strings that could be used, only that you're handling some known ones.

I'm less worried about whether `$this` and `$args` will be defined at all, and more concerned about the predictability of their value types. I don't think using a struct helps with that.

Reply:
If we're worried about variables needing to be statically checked against a known list (vs dynamic lookup in a hashmap) then maybe we should just have the variable extracted into its known enum variant during parsing? I'm assuming that we'd want to have users know at composition time that a top-level variable (a.k.a anything immediately touching the `$`) isn't defined.